### ISSUE TITLE
Timer stops emitting "timeout" after running for ~8 days

### ISSUE THREAD

**Issue by starryalley** [godotengine/godot#38878]
**Godot version:**
3.1.2 stable

**OS/device including version:**
Windows 10

**Issue description:**
I've built a godot project basically running static 2D image/slideshow on request. There are multiple threads written in gdscript there with UDP server handling requests to show images or slideshow.

After running the project for ~8 days, Timer seemed to stop working all of a sudden.

**Steps to reproduce:**

1.  everything was good until about keeping the godot project running ~8 days
2. when issue happened, a long running thread was doing this to play slideshow
```
         while active:
		emit_signal("update_image", ...) # it's a CONNECT_DEFERRED signal
		slideshow_cur_img += 1  # what will happen if this overflows?
		OS.delay_msec(1000 * slideshow_interval)
```
3. an existing timer suddenly stopped working. The `_on_timer_timeout` was never run again.
```
	timer = Timer.new()
	timer.set_wait_time(sync_interval_sec)
	timer.connect("timeout", self, "_on_timer_timeout")
	add_child(timer)
	timer.start()
```
4. new timer also stops working as well. The code below hung there.
```
        yield(get_tree().create_timer(wait_sec), "timeout")
```

While issue happened, all other parts of the godot project continued working.

**Minimal reproduction project:**
I don't have it now.

**Comment by aaronfranke** [comment]
Does your application happen to be running at 24 FPS? [I think I know the cause](https://www.google.com/search?q=%282%5E24+%2F+24%29+seconds+to+days).

#38880 should fix this, but I haven't tested it since that would take a long time.

**Comment by starryalley** [comment]
Thanks @aaronfranke . However I didn't set FPS specifically. But that fix makes sense!

To be specific, the starting time of my application was:
2020-05-01 17:35:39

And the last successful timer was triggered at:
2020-05-09 03:07:49

The timer wait_time was set to 3min and the next one never came. 

I will have a custom build myself and hopefully I can report back a few weeks later :P

**Comment by akien-mga** [comment]
I'm not sure #38880 would fix it, from what I understand @starryalley was not using a Timer with a wait time higher than `FLT_MAX` (~`3.4e+38`), but instantiating new Timers each time with a wait time likely within reasonable float bounds.

**Comment by Ev1lbl0w** [comment]
I think the problem is here, since `idle_process_time` it's still a float:https://github.com/godotengine/godot/blob/00949f0c5fcc6a4f8382a4a97d5591fd9ec380f8/scene/main/scene_tree.h#L255

**Comment by starryalley** [comment]
> I'm not sure #38880 would fix it, from what I understand @starryalley was not using a Timer with a wait time higher than `FLT_MAX` (~`3.4e+38`), but instantiating new Timers each time with a wait time likely within reasonable float bounds.

For more information, both an existing repeated Timer (instantiated at program start only once) and newly instantiated Timer stopped working. The wait time for the repeated timer is 3 min, and 1 second for the newly created one.

**Comment by aaronfranke** [comment]
@akien-mga 99.99% of the time, issues with floats aren't issues with being out of the absolute range, they are issues with being out of the accurate range. After 16777216, single-precision floats lose the ability to store all integers (16777215 -> 16777216 -> 16777218, since 16777217 is not a valid float), so if a program depends on incrementing by 1, it will fail at 16777216.

Similarly, if a program depends on incrementing by a smaller number, say, 1/24 per frame for 24 FPS, then it will fail earlier, approximately 24x as early, but in reality depends on how the number rounds to powers of two.

I don't know for sure if #38880 fixes the problem, but I know that using float for time is borked, so it should be replaced with double. I'll update the PR to also do this for SceneTree `get_idle_process_time()`.

**Comment by lawnjelly** [comment]
There's a lot of things going on in your description. This really needs a min production project (even though I accept one is hard to make, maybe just isolate one of the things that breaks in a single project). Then we can modify the code to pretend it has been running a long time.

This could simply be a QueryPerformanceCounter wraparound:
https://stackoverflow.com/questions/5296990/queryperformancecounter-and-overflows

**Comment by lawnjelly** [comment]
> I think the problem is here, since `idle_process_time` it's still a float:
> 
> https://github.com/godotengine/godot/blob/00949f0c5fcc6a4f8382a4a97d5591fd9ec380f8/scene/main/scene_tree.h#L255

`idle_process_time` being a float isn't necessarily a problem, provided it is promoted to a double in calculations where it needs to be a double. And of course provided that `idle_process_time` itself isn't being wrongly calculated....

**Comment by qarmin** [comment]
Can anyone check if it is fixed?
Related PR was already merged and cherrypicked.

**Comment by aaronfranke** [comment]
Closing as solved by #38880 and/or #51294 since nobody has reported timers bring broken in a long time.
